bundles/org.openhab.core.io.rest.update/src/main/java/org/openhab/core/io/rest/update/internal/updaters/BaseUpdater.java

```diff
@@ -0,0 +1,810 @@
+/**
+ * Copyright (c) 2010-2021 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.core.io.rest.update.internal.updaters;
+
+import java.io.BufferedReader;
+import java.io.ByteArrayOutputStream;
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.math.BigInteger;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Calendar;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.stream.Collectors;
+
+import javax.xml.stream.XMLInputFactory;
+import javax.xml.stream.XMLStreamConstants;
+import javax.xml.stream.XMLStreamException;
+import javax.xml.stream.XMLStreamReader;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.openhab.core.OpenHAB;
+import org.openhab.core.io.rest.update.internal.dto.OpenHabVersionDTO;
+import org.openhab.core.io.rest.update.internal.dto.UpdaterStatusDTO;
+import org.openhab.core.io.rest.update.internal.enums.TriState;
+import org.openhab.core.io.rest.update.internal.enums.VersionType;
+import org.osgi.framework.FrameworkUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.jcraft.jsch.ChannelExec;
+import com.jcraft.jsch.JSch;
+import com.jcraft.jsch.JSchException;
+import com.jcraft.jsch.Session;
+
+/**
+ * The {@link BaseUpdater} is a base for extending classes that will update openHAB.
+ *
+ * For each different operating system (resp. Linux package manager variety) there will be a different class that
+ * extends this class and executes the respective update process for the target system.
+ *
+ * @author AndrewFG - Initial contribution
+ */
+@NonNullByDefault
+public abstract class BaseUpdater implements Runnable {
+
+    protected final Logger logger = LoggerFactory.getLogger(BaseUpdater.class);
+
+    private static final String TITLE = "openHAB Updater";
+
+    private static final String ARTIFACT_URL_RELEASE = "https://openhab.jfrog.io/artifactory/libs-release-local/org/openhab/distro/openhab/maven-metadata.xml";
+    private static final String ARTIFACT_URL_MILESTONE = "https://openhab.jfrog.io/artifactory/libs-milestone-local/org/openhab/distro/openhab/maven-metadata.xml";
+    private static final String ARTIFACT_URL_SNAPSHOT = "https://openhab.jfrog.io/artifactory/libs-snapshot-local/org/openhab/distro/openhab/maven-metadata.xml";
```

Comment:
This will always only give you `-SNAPSHOT` versions as a result. If you'd then drill down another level to https://openhab.jfrog.io/artifactory/libs-snapshot-local/org/openhab/distro/openhab/3.1.0-SNAPSHOT/maven-metadata.xml, you could also read versions like `3.1.0-20210607.072644-149`, which would be much nicer for the user as they contain the exact timestamp of the build and would also allow to determine whether there is an update available or not (although it's not straight-forward to compare with the actual version as this has a different format, I admit).

Reply:
Thanks for the tip. Knowing the exact time stamp might eliminate the need for the “update available = don’t know” state mentioned elsewhere..

Reply:
Done.

Reply:
@kaikreuzer while investigating the drill down, I think I found a bug in the Artifactory: the maven-metadata shows three files, whereas in fact there are six files !! (see screenshots attached)..

![image](https://user-images.githubusercontent.com/893994/121229630-1b0de380-c886-11eb-9e1b-990ac320c5da.png)

![image](https://user-images.githubusercontent.com/893994/121229739-41cc1a00-c886-11eb-8260-dd7b7549cf0e.png)

Reply:
Nevertheless I have made the changes.